### PR TITLE
CXX-507 add bson examples

### PR DESCRIPTION
**Pull request by amidvidy**

### CODE REVIEW

File: examples/bsoncxx/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
+include_directories(
+    ${CMAKE_INSTALL_PREFIX}/${BSONCXX_HEADER_INSTALL_DIR}
+)
+
+link_directories(
+    ${CMAKE_INSTALL_PREFIX}/lib
+)
+
+set(BSONCXX_EXAMPLES
+  builder_core.cpp
```
Comment (amidvidy) [inline examples/bsoncxx/CMakeLists.txt]: indentation

File: examples/bsoncxx/builder_stream.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+#include <bsoncxx/builder/stream/array.hpp>
+#include <bsoncxx/builder/stream/helpers.hpp>
+#include <bsoncxx/types.hpp>
+
+using namespace bsoncxx;
+
+int main(int, char**) {
+
+
+    using builder::stream::document; // resolve some ambiguity
```
Comment (amidvidy) [inline examples/bsoncxx/builder_stream.cpp]: todo, remove this line comment as it is no longer neccessary

File: examples/bsoncxx/builder_basic.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    // basic::document builds a BSON document.
+    auto doc = builder::basic::document{};
+    // basic::array builds a BSON array
+    auto arr = builder::basic::array{};
+
+    // We append key-value pairs to a document using the kvp helper.
+    using bsoncxx::builder::basic::kvp;
+
+    doc.append(kvp("foo", "bar")); // string literal value will be converted to b_utf8 automatically
+    doc.append(kvp("baz", types::b_bool{false}));
```
Comment (acmorrow) [inline examples/bsoncxx/builder_basic.cpp]: I wonder if we should have b_false and b_true constants?

File: examples/bsoncxx/builder_core.cpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        builder.key_owned(temp);
+        builder.append(types::b_bool{false});
+    }
+
+    // 2. We can pass a key using key_view(). This is more efficient (less copying)
+    // but we need to manage the lifetime of the key until a matching value is appended, at which
+    // point it will be copied into an internally managed buffer.
+
+    {
+        // THIS IS ILLEGAL: DO NOT DO THIS
```
Comment (acmorrow) [inline examples/bsoncxx/builder_core.cpp]: I'm always a little hesitant about showing bad code, but I get what you are going for. It means, for instance, that if you run the examples under ASAN or valgrind they will emit errors.

File: examples/bsoncxx/builder_basic.cpp [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    // We can also pass a variable number of keys to append.
+    doc.append(kvp("a key", "a value"),
+               kvp("another key", "another value"),
+               kvp("moar keys", "moar values"));
+
+    // Appending to arrays is simple, just append one or more bson values.
+    arr.append("hello");
+    arr.append(types::b_bool{false}, types::b_bool{true}, types::b_double{1.234});
+
+    // If we want to create a subdocument, we can pass lambda as a value with a sub_document argument.
```
Comment (acmorrow) [inline examples/bsoncxx/builder_basic.cpp]: Looking at this, I can't remember whether this ends up doing in-place writing, or constructing a subdoc and then copying it. Maybe clarify in the comment?

File: examples/bsoncxx/view_and_value.cpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+               kvp("members", [](sub_array sa) {
+                   sa.append("tyler", "jason", "drew",
+                             "sam", "ernie", "john",
+                             "mark", "crystal");
+               })
+    );
+
+    // document::value is an owning bson document conceptually similar to string.
+    document::value value{doc.extract()};
+    // document::view is a non-owning bson document conceptually similar to string_view.
```
Comment (acmorrow) [inline examples/bsoncxx/view_and_value.cpp]: I think blank lines before the next comment would improve this

File: examples/bsoncxx/view_and_value.cpp [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    document::view view{value.view()};
+    // array::view and array::value are the corresponding classes for arrays.
+
+    // we can print a view using to_json
+    std::cout << to_json(view) << std::endl;
+
+    // note that all of the interesting methods for reading BSON are defined on the view type.
+
+    // iterate over the elements in a bson document
+    for (document::element ele : view) {
```
Comment (acmorrow) [inline examples/bsoncxx/view_and_value.cpp]: Can you say auto here, or did you want to say element explicitly to demonstrate the type?

File: examples/bsoncxx/view_and_value.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        // usually we don't need to actually use a switch statement, because we can also
+        // get a variant 'value' that can hold any BSON type.
+        types::value ele_val{ele.get_value()};
+        // if we need to print an arbitrary value, we can use to_json, which provides
+        // a suitable overload.
+        std::cout << "the value is " << to_json(ele_val) << std::endl;;
+    }
+
+    // If we want to search for an element we can use operator[]
+    // (we also provide a find() method that returns an iterator)
```
Comment (acmorrow) [inline examples/bsoncxx/view_and_value.cpp]: State that this is a linear search?

File: examples/bsoncxx/view_and_value.cpp [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    // Because view implements begin(), end(), we can also use standard STL algorithms.
+    // i.e. if we want to find the number of keys in a document we can use std::distance
+    using std::begin;
+    using std::end;
+
+    auto num_keys = std::distance(begin(view), end(view));
+    std::cout << "document has " << num_keys << " keys." << std::endl;
+    // i.e. if we want a vector of all the keys in a document, we can use std::transform
+    std::vector<std::string> doc_keys;
+    std::transform(begin(view), end(view), std::back_inserter(doc_keys), [](document::element ele) {
```
Comment (acmorrow) [inline examples/bsoncxx/view_and_value.cpp]: Neat!

File: examples/mongocxx/bson.cpp [hunk 10]
```diff
@@ -0,0 +1,16 @@
+#include <iostream>
```
Comment (acmorrow) [inline examples/mongocxx/bson.cpp]: Is this out of place, now?

**Comment by acmorrow** [comment]
LGTM, with a few nits and questions. I think this looks really nice.

File: examples/bsoncxx/view_and_value.cpp (hunk 7 above)
Comment (amidvidy) [inline examples/bsoncxx/view_and_value.cpp]: wanted to be as explicit as possible.

File: examples/mongocxx/bson.cpp (hunk 10 above)
Comment (amidvidy) [inline examples/mongocxx/bson.cpp]: yeah, this example should actually just be removed entirely.

File: examples/bsoncxx/builder_basic.cpp (hunk 3 above)
Comment (amidvidy) [inline examples/bsoncxx/builder_basic.cpp]: using native bool and false would actually work also - I just wanted to be clear about how data is really converted to a wrapped type.

File: examples/bsoncxx/builder_core.cpp (hunk 4 above)
Comment (amidvidy) [inline examples/bsoncxx/builder_core.cpp]: the builder itself goes out of scope here at the same time it becomes illegal to use, which should hopefully prevent ASAN from complaining.

**Comment by amidvidy** [comment]
cool, i responded to your comments. @TylerBrock, any thoughts on these?

File: examples/bsoncxx/view_and_value.cpp (hunk 8 above)
Comment (amidvidy) [inline examples/bsoncxx/view_and_value.cpp]: done

File: examples/bsoncxx/view_and_value.cpp [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+    auto num_keys = std::distance(begin(view), end(view));
+    std::cout << "document has " << num_keys << " keys." << std::endl;
+
+    // i.e. if we want a vector of all the keys in a document, we can use std::transform
+    std::vector<std::string> doc_keys;
+    std::transform(begin(view), end(view), std::back_inserter(doc_keys), [](document::element ele) {
+        // note that key() returns a string_view
+        return ele.key().to_string();
+    });;
```
Comment (amidvidy) [inline examples/bsoncxx/view_and_value.cpp]: extra semicolon here..

File: examples/bsoncxx/builder_basic.cpp (hunk 3 above)
Comment (TylerBrock) [inline examples/bsoncxx/builder_basic.cpp]: Should we split it into two examples then? An example with the native types in addition to one with explicit types might be nice but it also might break up the narrative.

File: examples/bsoncxx/builder_basic.cpp [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    auto arr = builder::basic::array{};
+
+    // We append key-value pairs to a document using the kvp helper.
+    using bsoncxx::builder::basic::kvp;
+
+    doc.append(kvp("foo", "bar")); // string literal value will be converted to b_utf8 automatically
+    doc.append(kvp("baz", types::b_bool{false}));
+    doc.append(kvp("garply", types::b_double{3.14159}));
+
+    // We can also pass a variable number of keys to append.
```
Comment (TylerBrock) [inline examples/bsoncxx/builder_basic.cpp]: Is there a limit to how many you can pass?

File: examples/bsoncxx/builder_basic.cpp [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    doc.append(kvp("a key", "a value"),
+               kvp("another key", "another value"),
+               kvp("moar keys", "moar values"));
+
+    // Appending to arrays is simple, just append one or more bson values.
+    arr.append("hello");
+    arr.append(false, types::b_bool{true}, types::b_double{1.234});
+
+    // If we want to create a subdocument, we can pass lambda as a value with a sub_document argument.
```
Comment (TylerBrock) [inline examples/bsoncxx/builder_basic.cpp]: Is this the only way? Maybe append arr to doc above to show that using a lambda is only an option?

File: examples/bsoncxx/builder_basic.cpp [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+             // subarrays work similarly
+             subarr.append(1, types::b_bool{false}, "hello", 5, [](sub_document subdoc) {
+                 // nesting works too!
+                 subdoc.append(kvp("such", "nesting"),
+                               kvp("much", "recurse"));
+             });
+         })
+    );
+
+    // We can get a view of the resulting bson by calling view()
```
Comment (TylerBrock) [inline examples/bsoncxx/builder_basic.cpp]: mention lifetime requirements of said view?

File: examples/bsoncxx/builder_core.cpp [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+using namespace bsoncxx;
+
+int main(int, char**) {
+
+    // bsoncxx::builder::core is a low-level primitive that can be useful for building other
+    // BSON abstractions. Most users should just use builder::stream or builder::basic.
+
+    // the boolean argument to core should be true if the top-level BSON datum is an array.
+    // if false, it will be created as a document.
+    auto builder = builder::core{false};
```
Comment (TylerBrock) [inline examples/bsoncxx/builder_core.cpp]: Ugh, boolean parameter at the call site. Good thing we document this.

File: examples/bsoncxx/builder_stream.cpp [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        << "baz" << types::b_int32{1234}
+        << "quz" << types::b_double{1.234};
+
+    // For arrays, everything just becomes a value
+    arr << 1 << 2 << types::b_bool{true};
+
+    // The stream namespace includes some helpers that can be used similarly
+    // to the stream manipulators in <iomanip>
+    // To build a subdocument, use open_document and close_document
+    using builder::stream::open_document;
```
Comment (TylerBrock) [inline examples/bsoncxx/builder_stream.cpp]: Include JSON version of the constructed document in comments for visual reference?

File: src/bsoncxx/CMakeLists.txt [hunk 17]
```diff
@@ -65,6 +65,8 @@ add_library(bsoncxx SHARED
     ${bsoncxx_sources}
 )
 
+set(bsoncxx_libs ${LIBBSON_LIBRARIES})
+
```
Comment (TylerBrock) [inline src/bsoncxx/CMakeLists.txt]: unrelated change?

**Comment by TylerBrock** [comment]
This looks hawt :fire:

File: examples/bsoncxx/builder_core.cpp (hunk 15 above)
Comment (acmorrow) [inline examples/bsoncxx/builder_core.cpp]: Actually, I mean to point this out too. Thanks for raising this @TylerBrock . Should we make this an enum or something? (In another commit, open a ticket, etc.).

File: examples/bsoncxx/builder_basic.cpp (hunk 12 above)
Comment (amidvidy) [inline examples/bsoncxx/builder_basic.cpp]: probably the C++ template instantiation limit, which is more than anyone will ever type.

File: src/bsoncxx/CMakeLists.txt (hunk 17 above)
Comment (amidvidy) [inline src/bsoncxx/CMakeLists.txt]: intentional, because I want to link the bsoncxx examples against the libs that bsoncxx links against. I figured it was good to put them in a var in case we ever add more.
